presto-hudi/src/main/java/com/facebook/presto/hudi/HudiFileSkippingManager.java

```diff
@@ -0,0 +1,312 @@
+        }
+        else {
+            allInputFileSlices = partitions.stream().collect(Collectors.toMap(Function.identity(), partition ->
+                    queryInstant.map(instant ->
+                            fileSystemView.getLatestFileSlicesBeforeOrOn(partition, instant, true))
+                            .orElse(fileSystemView.getLatestFileSlices(partition))
+                            .collect(Collectors.toList())));
+        }
+
+        long duration = System.currentTimeMillis() - startTime;
+
+        log.info(String.format("prepare query files for table %s, spent: %d ms", metaClient.getTableConfig().getTableName(), duration));
+    }
+
+    public Map<String, List<FileSlice>> listQueryFiles(TupleDomain<ColumnHandle> tupleDomain)
+    {
+        // do file skipping by MetadataTable
+        Map<String, List<FileSlice>> candidateFileSlices = allInputFileSlices;
+        try {
+            if (!tupleDomain.isAll()) {
+                candidateFileSlices = lookupCandidateFilesInMetadataTable(candidateFileSlices, tupleDomain);
+            }
+        }
+        catch (Exception e) {
+            // Should not throw exception, just log this Exception.
+            log.warn(String.format("failed to do data skipping for table: %s, fallback to all files scan", metaClient.getBasePathV2().toString()), e);
+            candidateFileSlices = allInputFileSlices;
+        }
+        if (log.isDebugEnabled()) {
+            int candidateFileSize = candidateFileSlices.values().stream().map(List::size).reduce(0, Integer::sum);
+            int totalFiles = allInputFileSlices.values().stream().map(List::size).reduce(0, Integer::sum);
+            double skippingPercent = totalFiles == 0 ? 0.0d : (totalFiles - candidateFileSize) / (totalFiles + 0.0d);
+            log.debug(String.format("Total files: %s; candidate files after data skipping: %s; skipping percent %s", totalFiles, candidateFileSize, skippingPercent));
+        }
+        return candidateFileSlices;
+    }
+
+    private Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(Map<String, List<FileSlice>> inputFileSlices, TupleDomain<ColumnHandle> tupleDomain)
+    {
+        // split regular column predicates
+        TupleDomain<HudiColumnHandle> regularTupleDomain = HudiPredicates.from(tupleDomain).getRegularColumnPredicates();
+        TupleDomain<String> regularColumnPredicates = regularTupleDomain.transform(HudiColumnHandle::getName);
+        if (regularColumnPredicates.isAll() || !regularColumnPredicates.getDomains().isPresent()) {
+            return inputFileSlices;
+        }
+        List<String> regularColumns = regularColumnPredicates.getDomains().get().entrySet().stream().map(Map.Entry::getKey).collect(Collectors.toList());
+        // get filter columns
+        List<String> encodedTargetColumnNames = regularColumns.stream().map(col -> new ColumnIndexID(col).asBase64EncodedString()).collect(Collectors.toList());
```

Comment:
may be in a follow up PR. we should also wire in pruned list of partitions here, so that we prefix look up only in pruned partitions rather than all partitions. For eg, if there are 1000 partitions and 100 cols, and only 10 partitions are matched after pruning, 

exiting call will fetch 100 cols * 1000 partitions = 10k entries to do file skipping. 
where as if we wire in pruned list of partitions, then we only need to do file skipping from 100 entries.

Reply:
guess we missed this even for spark impl in Hudi. will file a jira on this.

Reply:
https://issues.apache.org/jira/browse/HUDI-5245